### PR TITLE
Backend recipe selection and grocery list generation

### PR DESCRIPTION
**Pull request by Brianna-Flo**
## Description

<what this pull request is doing>

- enable a user to store recipes they want to create and retrieve when navigating to the "create new list" page
- Display recipe instructions in a modal upon click
- Allow users to delete recipes they do not want to include anymore
- set up fetch from walmart api (for pricing)
- Generate a grocery list with quantities needed (based on ingredients on hand) and combine recipe ingredients
- Sort grocery list by department

<what will be done in later PRs and not included here>

- improved error handing in conversion
- Add price estimation

## Milestones
<the milestones or stories/features that this works towards>

- backend functionality to generate a grocery list for a user based on recipes chosen

## Resources
<links to tutorials, code snippets, inspirations>
<if you took or translated specific code from the internet, please call it out here!>
Prisma: https://www.prisma.io/docs/orm/prisma-client/special-fields-and-types/working-with-json-fields
Walmart API setup: https://metaschool.so/articles/node-js-crypto-createsign-method
Walmart API setup: https://metaschool.so/articles/node-js-crypto-createsign-method
Walmart API: https://walmart.io/docs/walmart-identity/v1/supporting-information
Walmart API: https://walmart.io/docs/affiliates/v1/search
Unit conversion: https://www.npmjs.com/package/convert-units
Array .find: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find
Array .filter: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter
Dictionary: https://graphite.dev/guides/typescript-dictionary
Record type: https://graphite.dev/guides/typescript-record-utility-type

## Test Plan
<insert images or gifs of feature>
<otherwise, say how code was tested if not UI facing>
<any edge cases you might have specifically tested>

- tested recipes that share common ingredients (combined into a single grocery entry)
- tested unit conversion to have single quantity needed for common ingredients
- If user has enough of ingredient, not presented in grocery list
- if user doesnt have ingredient, presented in grocery list
- if user has ingredient but not enough, present only amount needed in grocery list
<div>
    <a href="https://www.loom.com/share/5d3fd73edf55415380aed6fd656c0edb">
      <p>Generate grocery list based on ingredients on hand - Watch Video</p>
    </a>
    <a href="https://www.loom.com/share/5d3fd73edf55415380aed6fd656c0edb">
      
    </a>
  </div>

### CODE REVIEW

File: frontend/src/pages/Homepage.tsx [hunk 1]
```diff
@@ -1,14 +1,26 @@
...
+  const [message, setMessage] = useState<GPErrorMessageTypes>()
+  const [userGroceryList, setUserGroceryList] = useState<GPRecipeIngredientTypes[]>([])
+  const [userIngredientList, setUserIngredientList] = useState<GPIngredientDataTypes[]>([])
+  useEffect(() => {
+    const setLists = async () => {
+      fetchGroceryList({setMessage, setUserGroceryList});
+      const userIngredients = await fetchUserIngredientsHelper({setMessage})
+      setUserIngredientList(userIngredients)
+    }
+    setLists();
```
Comment (nithyad) [inline frontend/src/pages/Homepage.tsx:22]: Maybe use a more creative name with this than setLists

File: frontend/src/utils/databaseHelpers.ts [hunk 2]
```diff
@@ -300,5 +304,5 @@ export {
   updateIngredientDatabase,
   fetchRecipes,
   updateUserRecipes,
-  fetchGroceryList
+  fetchGroceryList,
```
Comment (nithyad) [inline frontend/src/utils/databaseHelpers.ts:307]: Is there a reason this comma was added?

**Review by nithyad — APPROVED** [review]

**Review by nithyad — APPROVED** [review]
LGTM!

File: backend/utils/constants.ts [hunk 3]
```diff
@@ -1,65 +1,65 @@
...
+  kilograms: "kg",
 
-    oz: 'oz', 
-    "oz.": 'oz', 
-    ounce: 'oz', 
-    ounces: 'oz', 
+  oz: "oz",
+  "oz.": "oz",
+  ounce: "oz",
+  ounces: "oz",
```
Comment (nithyad) [inline backend/utils/constants.ts:21]: same here with capitalization

File: frontend/src/pages/Homepage.tsx (hunk 1 above)
Comment (Brianna-Flo) [reply frontend/src/pages/Homepage.tsx:22]: Would setUserListPreviews be a better alternative?

File: frontend/src/utils/databaseHelpers.ts (hunk 2 above)
Comment (Brianna-Flo) [reply frontend/src/utils/databaseHelpers.ts:307]: That may be a result of the auto formatting from Prettier

File: backend/utils/constants.ts [hunk 4]
```diff
@@ -1,65 +1,65 @@
...
+  grams: "g",
 
-    kg: 'kg', 
-    'kg.': 'kg', 
-    kilogram: 'kg', 
-    kilograms: 'kg', 
+  kg: "kg",
+  "kg.": "kg",
+  kilogram: "kg",
+  kilograms: "kg",
```
[the remark this answers is not in the document]
Comment (Brianna-Flo) [reply backend/utils/constants.ts:16]: The unit conversions library has set formats for the unit strings passed, and Tsp is the only one capitalized. Prior to passing the unit into the dictionary I convert the string to lowercase. I deleted the capitalized Tbsp. to avoid confusion.

File: backend/prisma/migrations/20250705180439_ingredients_optional/migration.sql [hunk 5]
```diff
@@ -0,0 +1,2 @@
+-- AlterTable
+ALTER TABLE "Ingredient" ALTER COLUMN "expiration" SET DATA TYPE TEXT;
```
Comment (radhikadhomse) [inline backend/prisma/migrations/20250705180439_ingredients_optional/migration.sql:2]: curious why we want to use a text instead of a numerical timestamp 
will we ever need to calcualte anything?

File: backend/prisma/migrations/20250706044953_ingredients_on_hand/migration.sql [hunk 6]
```diff
@@ -0,0 +1,13 @@
+-- CreateTable
+CREATE TABLE "IngredientsOnHand" (
```
Comment (radhikadhomse) [inline backend/prisma/migrations/20250706044953_ingredients_on_hand/migration.sql:2]: this is a nit so totally up to you -- can we do something like OwnedIngredients or XIngredients that conveys the same thing

File: backend/prisma/schema.prisma [hunk 7]
```diff
@@ -14,69 +14,55 @@ datasource db {
...
+  intolerances      String[]
   diets             String[]
-  ingredientsOnHand Ingredient[]
-  groceryList       GroceryList? // store the grocery list for a user
+  ingredientsOnHand IngredientsOnHand[]
+  groceryList       Json?
+  recipes           Recipe[]
+}
+
+model IngredientsOnHand {
```
Comment (radhikadhomse) [inline backend/prisma/schema.prisma:28]: probably should be IngredientOnHand rather than IngredientsOnhand becasue each represents one ingredient and the quantity of it

File: backend/prisma/schema.prisma [hunk 8]
```diff
@@ -14,69 +14,55 @@ datasource db {
...
-  id        Int          @id @default(autoincrement())
-  listName  String
-  startDate DateTime
-  endDate   DateTime
-  toBuy     Ingredient[]
-  userId    Int          @unique
-  user      User         @relation(fields: [userId], references: [id])
+  id           Int      @id @default(autoincrement()) // store the id number from spoonacular api, otherwise id is auto incremented (part of the users database)
+  apiId        Int      @unique
+  recipeTitle  String
```
Comment (radhikadhomse) [inline backend/prisma/schema.prisma:56]: what if you have two recipes called "Chia Pudding"?

File: backend/utils/utils.ts [hunk 9]
```diff
@@ -0,0 +1,174 @@
+const { PrismaClient } = require("@prisma/client");
```
Comment (radhikadhomse) [inline backend/utils/utils.ts:1]: import { PrismaClient } from "@prisma/client";


just for consistency

File: backend/utils/utils.ts [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+};
+
+type GPConvertUnitsType = {
+  convertTo: GPRecipeIngredientTypes;
+  converting: GPRecipeIngredientTypes;
+};
+const convertUnits = ({ convertTo, converting }: GPConvertUnitsType) => {
+  let convertToUnit = unitConversions[convertTo.unit.toLowerCase()];
+  let convertingUnit = unitConversions[converting.unit.toLowerCase()];
+  let convertingQuantity = parseInt(converting.quantity);
```
Comment (radhikadhomse) [inline backend/utils/utils.ts:25]: is float better?

File: backend/utils/utils.ts [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+  recipeIngredient,
+  recipeIngredients,
+}: GPTotalQuantityTypes) => {
+  // for the current ingredient, check to see if there are any of the same ingredient in the list
+  const sameIngredients = recipeIngredients.filter(
+    (ingredient) =>
+      ingredient.ingredientName === recipeIngredient.ingredientName
+  );
+  // loop through array of ingredients to get the quantity needed
+  let totalQuantity = 0;
```
Comment (radhikadhomse) [inline backend/utils/utils.ts:60]: if there’s only one matching ingredient, you return 0

File: frontend/src/App.tsx [hunk 12]
```diff
@@ -14,10 +14,16 @@ import NoMatchPage from "./pages/NoMatchPage";
 import GroceryList from "./pages/GroceryList";
 import SignupForm from "./pages/SignupForm";
 import AccountPage from "./pages/AccountPage";
+import WithAuth from "./components/WithAuth";
 
 function App() {
   const [navOpen, setNavOpen] = useState<boolean>(false);
 
+  const ProtectedNewListPage = WithAuth(NewListPage);
```
Comment (radhikadhomse) [inline frontend/src/App.tsx:22]: can we just do AuthNewListPage

**Review by radhikadhomse — CHANGES_REQUESTED** [review]
comments separate
